cmdstanpy/lib.py

```diff
@@ -527,27 +552,59 @@ def sample(self) -> np.ndarray:
         all draws from a chain are contiguous.
         """
         if self._sample is None:
-            self._sample = self.get_sample()
+            self.parse_sample()
         return self._sample
 
-    def get_sample(self) -> np.ndarray:
+    def parse_sample(self) -> None:
         """
-        Returns posterior sample.
-        The first time this function is called it assembles the sample
-        from the stan_csv files; subsequent calls to this function
-        return the assembled sample.
+        Allocates and populates the stepsize, metric, and drawset arrays
+        by parsing the validated stan_csv files if they don't already exist.
         """
-        sample = np.empty(
+        if not (self._stepsize is None and self._metric is None and
+                    self._sample is None):
+            return
+        self._stepsize = np.empty(self._chains, dtype=float)
+        if self._metric_type == 'diag_e':
+            self._metric = np.empty(
+                (self._chains, self._num_params), dtype=float)
+        else:
+            self._metric = np.empty(
+                (self._chains, self._num_params, self._num_params),
+                dtype=float)
+        self._sample = np.empty(
             (self._draws, self._chains, len(self._column_names)), dtype=float,
             order='F'
         )
         for chain in range(self._chains):
-            draw = 0
-            with open(self._csv_files[chain], 'r') as fd:
-                for line in fd:
-                    if line.startswith('#') or line.startswith('lp__,'):
-                        continue
-                    vs = [float(x) for x in line.split(',')]
-                    sample[draw, chain, :] = vs
-                    draw += 1
-        return sample
+            try:
+                fp = open(self._csv_files[chain], 'r')
```

Comment:
This could be handled with `with` that does close the file at the end

Reply:
didn't realize that 'with' statement builds in try/except/finally behavoir.  thanks for the correction!